test/EFCore.Specification.Tests/Query/ComplexNavigationsQueryTestBase.cs

```
@@ -4980,5 +4980,284 @@ public virtual Task Contains_over_optional_navigation_with_null_entity_reference
+                .Include(l1 => l1.OneToMany_Optional1.Where(x => ctx.LevelOne.Count(xx => xx.Id != x.Id) > 1).OrderBy(x => x.Id).Take(3));
+            var result = query.ToList();
+        }
+
+        [ConditionalTheory]
+        [MemberData(nameof(IsAsyncData))]
+        public virtual async Task Filtered_include_include_parameter_used_inside_filter_throws(bool async)
+        {
+            await Assert.ThrowsAsync<InvalidOperationException>(
+                () => AssertQuery(
+                    async,
+                    ss => ss.Set<Level1>()
+                        .Select(l1 => ss.Set<Level2>().Include(l2 => l2.OneToMany_Optional2.Where(x => x.Id != l2.Id)))));
+        }
+
+        [ConditionalFact]
+        public virtual void Filtered_include_outer_parameter_used_inside_filter()
+        {
+            // TODO: needs #18191 for result verification
+            using var ctx = CreateContext();
+            var query = ctx.LevelOne.Select(l1 => new
+            {
+                l1.Id,
+                FullInclude = ctx.LevelTwo.Include(l2 => l2.OneToMany_Optional2).ToList(),
+                FilteredInclude = ctx.LevelTwo.Include(l2 => l2.OneToMany_Optional2.Where(x => x.Id != l1.Id)).ToList() });
+            var result = query.ToList();
+        }
+
+        [ConditionalFact]
+        public virtual void Filtered_include_is_considered_loaded()
+        {
+            using var ctx = CreateContext();
+            var query = ctx.LevelOne.AsTracking().Include(l1 => l1.OneToMany_Optional1.OrderBy(x => x.Id).Take(1));
+            var result = query.ToList();
+            foreach (var resultElement in result)
+            {
+                var entry = ctx.Entry(resultElement);
+                Assert.True(entry.Navigation("OneToMany_Optional1").IsLoaded);
+            }
+        }
+
+        [ConditionalTheory]
+        [MemberData(nameof(IsAsyncData))]
+        public virtual async Task Filtered_include_with_Distinct_throws(bool async)
```

Add a test where navigation is registered as reference navigation in EF Core metadata but implements IEnumerable<> and will allow you to write filtered include syntax. (Essentially filterExpression when put on non collection navigation should throw.

---

Do people actually do that? Seems like very contrived scenario.

---

It's negative case example. Rather than ignoring (which what we are doing currently when we see non-navigation in Include, we should change that), we can throw a better error message.

---

On note of reference nav which implements IEnumerable, it is not impossible.